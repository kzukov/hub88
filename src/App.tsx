import { useQuery } from "@apollo/client";
import { useState } from "react";
import Alert from "./components/alert/alert";
import Container from "./components/container/container";
import Input from "./components/input/input";
import Row from "./components/row/row";
import Table, { TableColumn } from "./components/table/table";
import TopSpacing from "./components/top-spacing/top-spacing";
import {
  BY_COUNTRY_CODE,
  ByCountryCode,
  CountryVariables,
} from "./queries/by-country-code";

function App() {
  const [countryCode, setCountryCode] = useState("");

  const { data, loading, error } = useQuery<ByCountryCode, CountryVariables>(
    BY_COUNTRY_CODE,
    { variables: { code: countryCode } }
  );

  const columns: TableColumn[] = [
    {
      key: "name",
      content: "Country name",
    },
    {
      key: "code",
      content: "Country code",
    },
  ];

  return (
    <Container>
      <Row>
        <TopSpacing>
          <Alert type="warning">Example input: EE</Alert>
          <TopSpacing>
            <Input
              placeholder="Filter by country code"
              onChange={(value) => {
                setCountryCode(value);
              }}
            />
          </TopSpacing>
          <hr />
          {error && (
            <TopSpacing>
              <Alert type="danger">{error.toString()}</Alert>
            </TopSpacing>
          )}
          {data?.country && (
            <TopSpacing>
              <Table columns={columns} data={[data.country]} />
            </TopSpacing>
          )}
          {!data?.country && (
            <TopSpacing>
              <Alert type="dark">No data</Alert>
            </TopSpacing>
          )}
          {loading && (
            <TopSpacing>
              <Alert type="primary">Data is loading</Alert>
            </TopSpacing>
          )}
        </TopSpacing>
      </Row>
    </Container>
  );
}

export default App;
