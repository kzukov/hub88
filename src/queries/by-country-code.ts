import { gql } from "@apollo/client";

export interface ByCountryCode {
  country: {
    code: string;
    name: string;
  };
}

export interface CountryVariables {
  code: string;
}

export const BY_COUNTRY_CODE = gql`
  query ByCountryCode($code: ID!) {
    country(code: $code) {
      name
      code
    }
  }
`;
