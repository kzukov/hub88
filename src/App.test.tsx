import React from "react";
import App from "./App";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MockedProvider } from "@apollo/client/testing";
import { BY_COUNTRY_CODE } from "./queries/by-country-code";

describe("<App />", () => {
  test("matches snapshot", () => {
    const { asFragment } = render(
      <MockedProvider addTypename={false}>
        <App />
      </MockedProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("makes successfull request", async () => {
    const mocks = [
      {
        request: {
          query: BY_COUNTRY_CODE,
          variables: {
            code: "EE",
          },
        },
        result: {
          data: {
            country: { code: "EE", name: "Estonia" },
          },
        },
      },
    ];

    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <App />
      </MockedProvider>
    );

    const input = screen.getByRole<HTMLInputElement>("textbox");

    fireEvent.change(input, { target: { value: "EE" } });

    expect(await screen.findByText("Estonia")).toBeInTheDocument();
  });

  test("makes failed request", async () => {
    const mocks = [
      {
        request: {
          query: BY_COUNTRY_CODE,
          variables: {
            code: "EE",
          },
        },
        error: new Error("An error occurred"),
      },
    ];

    render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <App />
      </MockedProvider>
    );

    const input = screen.getByRole<HTMLInputElement>("textbox");

    fireEvent.change(input, { target: { value: "EE" } });

    expect(
      await screen.findByText("Error: An error occurred")
    ).toBeInTheDocument();
  });
});
