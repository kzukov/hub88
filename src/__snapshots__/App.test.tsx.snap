// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="mt-3"
      >
        <div
          class="alert alert-warning"
          role="alert"
        >
          Example input: EE
        </div>
        <div
          class="mt-3"
        >
          <input
            class="form-control"
            placeholder="Filter by country code"
            type="text"
          />
        </div>
        <hr />
        <div
          class="mt-3"
        >
          <div
            class="alert alert-dark"
            role="alert"
          >
            No data
          </div>
        </div>
        <div
          class="mt-3"
        >
          <div
            class="alert alert-primary"
            role="alert"
          >
            Data is loading
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
